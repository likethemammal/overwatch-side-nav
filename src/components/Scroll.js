import React, { Component } from 'react'

import _ from 'underscore'

import {
    isAtBottomOfPage,
} from '../units'

import {
    lastHash as _lastHash,
    newHash as _newHash,
    hasHashChanged as _hasHashChanged,
} from '../selectors'

const BOTTOM_SCROLL_PADDING = 150

class Scroll extends Component {

    static defaultProps = {
        bottomPadding: BOTTOM_SCROLL_PADDING,
    }

    constructor() {
        super()

        this.onScroll = _.throttle(this._onScroll, 100)
    }

    _onScroll = () => {

        const { onHashChange, bottomPadding } = this.props

        const lastHash = _lastHash(this.props)
        const isAtBottom = isAtBottomOfPage(bottomPadding)

        if (isAtBottom) {
            onHashChange(lastHash)
            return
        }

        const newHash = _newHash(this.props)
        const hasHashChanged = _hasHashChanged(this.props)

        if (hasHashChanged) {
            onHashChange(newHash)
        }
    }

    componentDidMount() {
        window.addEventListener('scroll', this.onScroll)
    }

    componentWillUnmount() {
        window.removeEventListener('scroll', this.onScroll)
    }

    render() {
        return <div/>
    }
}

export default Scroll