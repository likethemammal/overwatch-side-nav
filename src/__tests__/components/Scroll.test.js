import React from 'react';
import {shallow} from 'enzyme';

import general from '@likethemammal/general-components'

const shared = general.__tests__._shared.tests

import Scroll from '../../components/Scroll'

const mock_onHashChange = jest.fn()

const items = [
    {
        label: 'Label',
        id: 'some_id',
        icon: 'svg-file-stub',
    },
    {
        label: 'Test',
        id: 'slug',
        icon: 'svg-file-stub',
    },
]


describe('Scroll', () => {

    const component = shallow(
        <Scroll
            items={items}
            onHashChange={mock_onHashChange}
        />
    )

    shared.SNAPSHOTS_SHOULD_MATCH([
        component
    ])

})