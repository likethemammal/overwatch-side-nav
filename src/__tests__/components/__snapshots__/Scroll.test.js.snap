// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scroll Snapshot 0 should match 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Scroll
    items={
      Array [
        Object {
          "icon": "svg-file-stub",
          "id": "some_id",
          "label": "Label",
        },
        Object {
          "icon": "svg-file-stub",
          "id": "slug",
          "label": "Test",
        },
      ]
    }
    onHashChange={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
